### DISCUSSION TITLE
[Feature] Node Graph Control

### DISCUSSION THREAD

**Discussion by Avid29** [CommunityToolkit/Labs-Windows#235] [Ideas]
## Describe the problem this feature would solve

A node graph is a fairly niche control, but as UWP becomes used for more technical projects flow charts are more common. This would be a useful control for creating a graphs visually.

## Describe the solution

A control with Nodes, Ports, and Connections. Similar to Blender, Unity, or UE4 shaders. 

## Describe alternatives you've considered

Anytime someone wants this control, they would need to write it themselves. 

## Additional context & Screenshots

![image](https://user-images.githubusercontent.com/9384045/144917256-bc908b7e-bdf4-4a55-9a96-0337f27d6142.png)

**Comment by ghost** [comment]
Hello, 'Avid29! Thanks for submitting a new feature request. I've automatically added a vote 👍 reaction to help get things started. Other community members can vote to help us prioritize this feature in the future!

**Comment by Arlodotexe** [comment]
cc @djspider117 you may be interested in collaborating on this.

**Comment by djspider117** [comment]
I have a prototype that is native UWP 
![image](https://user-images.githubusercontent.com/20155221/146224695-c1b29a36-6739-4d57-86d1-afd9eab866a4.png)

but it uses win2d (for drawing the lines) and i wonder if that is a restriction

**Comment by Arlodotexe** [comment]
> but it uses win2d (for drawing the lines) and i wonder if that is a restriction

It shouldn't be an issue, but it does make it harder to port x-plat. If being done from scratch, maybe we could try porting that part to SkiaSharp?

**Comment by Avid29** [comment]
> but it uses win2d (for drawing the lines) and i wonder if that is a restriction

In my opinion the curves should also be a templated control. Is it not possible to just use a `<Path/>`?

**Comment by Arlodotexe** [comment]
If it can't be a templated control, the style should at least be customizable.

**Comment by Difegue** [comment]
+1 on this, would be open to contributing to a control like that as I have potential needs for one 😌  

I think using Win2D for the paths would be fine for a first version of the control (You'd already have a fair bit of templating to do for the nodes themselves), I also don't think any controls in the WCT currently use skiasharp?   
Do correct me if I'm wrong on this 😅

**Comment by Avid29** [comment]
Someday, if nobody else makes this I will. In the meantime, I did some research a while ago and I'm willing to share what I learned with anyone who's interested. 

Looks like Difegue made it further though

**Comment by Difegue** [comment]
Oh sorry to disappoint but I didn't 😂 Just saw this issue while doing some searching.  
I'm likely going to fallback to using https://github.com/panthernet/GraphX for my own project, but a simpler toolkit control would've probably been better/leaner.

**Comment by lukedukeus** [comment]
+1 on this too, https://github.com/miroiu/nodify is super cool, but only supports wpf ☹️  A uwp/ winui implementation would be cool
